### PR TITLE
Continuum BG Estimation 

### PR DESCRIPTION
**Pull request by ckarwin**
This is an initial method for estimating the background for continuum sources. It is still highly preliminary, and so I am opening this as a PR draft, in order to share the example notebook which has the initial algorithm hard-coded (available [here](https://github.com/ckarwin/cosipy/blob/develop/docs/tutorials/background_estimation/continuum_estimation/continuum_BG_estimation_example.ipynb)). I plan to develop the `ContinuumEstimation` class leading up to the October meeting, in order to at least have a working version that can then be further develop. I think the `ContinuumEstimation` class should be part of a `background_estimation` base class, which will also contain a `LineEstimation` class and a `BurstEstimation` class. But we should discuss this and figure out what makes the most sense. 

Here are slides summarizing the initial results from the method presented in the example tutorial: 
[continuum_bg_estimation.pdf](https://github.com/user-attachments/files/16698357/continuum_bg_estimation.pdf). Below is a summary of the next steps:

**Improving algorithm:**

- Optimize cut for mask selection (so far only tested cumdist < 0.4)
- Use actual ARM measurement for mask selection
- Use more sophisticate inpainting method:

     - Inpainting Galactic Foreground Intensity and Polarization Maps Using Convolution Neural Networks (Puglisi+20, [link](https://iopscience.iop.org/article/10.3847/1538-4357/abc47c))
          - Tested for small angular scales, but what about large scales?
     - SMILE inpainting algorithm: (Li 2014, [link](https://haifengl.github.io/))
          - Based on solving the Laplace equation for each missing pixel by using the values of the surrounding pixels. Might not work so well for our purposes. 
          - Also used in Macias+19 ([link](https://iopscience.iop.org/article/10.1088/1475-7516/2019/09/042)). 
     - We used CNN for a somewhat similar problem: Deep learning models of the discrete component of the Galactic interstellar gamma-ray emission ([link](https://journals.aps.org/prd/abstract/10.1103/PhysRevD.107.063018)).
     - Develop our own, optimized for our specific problem (maybe employing some ML algorithm). 
     - Other methods?

- Give more freedom to fit (first need to fix unrelated convergence problem).
- Vectorize code

**More tests:**

- Test method on extended/diffuse source.
- Test Crab fit with GDCE included in total BG.
- Test Crab fit with other point source(s) included in BG. 
- Test Crab fit will all other astrophysical sources included in BG.

### CODE REVIEW

File: cosipy/background_estimation/ContinuumEstimation.py [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+            Full path to precomputed response file (.h5 file).
+        """
+
+        logger.info("...loading the pre-computed image response ...")
+        self.psr = DetectorResponse.open(psr_file)
+        logger.info("--> done")
+
+        return
+
+    def laod_full_data(self, data_file, data_yaml):
```
Comment (hiyoneda) [inline cosipy/background_estimation/ContinuumEstimation.py:73]: load (typo?)

Comment (ckarwin) [reply cosipy/background_estimation/ContinuumEstimation.py]: Yup, thanks! Fixed now.

**Review by hiyoneda — COMMENTED** [review]
I have reviewed the code. I added some suggestions about the code.

File: cosipy/background_estimation/ContinuumEstimation.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+import numpy as np
+from scipy.stats import norm
+import numpy.ma as ma
+from tqdm import tqdm
+import logging
+logger = logging.getLogger(__name__)
+
+class ContinuumEstimation:
+    
+    def calc_psr(self, ori_file, detector_response, coord, output_file, nside=16):
```
[the remark this answers is not in the document]
Comment (ckarwin) [reply cosipy/background_estimation/ContinuumEstimation.py:19]: Yes, that makes sense.

File: cosipy/background_estimation/ContinuumEstimation.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+
+        # Scatt map:
+        scatt_map = sc_orientation.get_scatt_map(coord, nside = nside, coordsys = 'galactic')
+
+        # Calculate PSR:
+        with FullDetectorResponse.open(dr) as response:
+            self.psr = response.get_point_source_response(coord = coord, scatt_map = scatt_map)
+
+        # Save:
+        self.psr.write(output_file + ".h5")
```
[the remark this answers is not in the document]
Comment (ckarwin) [reply cosipy/background_estimation/ContinuumEstimation.py:51]: Yes, done.

File: cosipy/background_estimation/ContinuumEstimation.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+                if next_value != 0:
+                    interp_list_low.append(next_value)
+                    search_left = False
+           
+            # Search right:
+            j = 1
+            search_right = True
+            while search_right == True:
+               
+                if this_index+j >= len(self.psr.axes['PsiChi'].centers)-1:
```
[the remark this answers is not in the document]
Comment (ckarwin) [reply cosipy/background_estimation/ContinuumEstimation.py:208]: More elegant, thanks! Changed.

File: cosipy/background_estimation/ContinuumEstimation.py [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+        if make_plots == True:
+            plt.plot(cumdist)
+            plt.title("Cumulative Distribution")
+            plt.xlabel("Pixel")
+            plt.ylabel("Fraction of Counts")
+            plt.savefig("cumdist.png")
+            plt.show()
+            plt.close()
+        
+        return
```
[the remark this answers is not in the document]
Comment (ckarwin) [reply cosipy/background_estimation/ContinuumEstimation.py:149]: I made your suggested changes.

File: cosipy/background_estimation/ContinuumEstimation.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+            plt.title("Cumulative Distribution")
+            plt.xlabel("Pixel")
+            plt.ylabel("Fraction of Counts")
+            plt.savefig("cumdist.png")
+            plt.show()
+            plt.close()
+        
+        return 
+
+    def simple_inpainting(self, m_data):
```
[the remark this answers is not in the document]
Comment (ckarwin) [reply cosipy/background_estimation/ContinuumEstimation.py:151]: Done.

File: cosipy/background_estimation/ContinuumEstimation.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+      
+        # Close progress bar:
+        pbar.close()
+
+        # Write estimated BG file:
+        logger.info("Writing file...")
+        self.estimated_bg.write(output_file,overwrite=True)
+        logger.info("Finished!")
+
+        return
```
[the remark this answers is not in the document]
Comment (ckarwin) [reply cosipy/background_estimation/ContinuumEstimation.py:354]: Done.

File: cosipy/background_estimation/ContinuumEstimation.py [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+        -----
+        In practice, the data file used for estimating the background 
+        should be the full dataset.  
+        
+        The full data binning needs to match the PSR. 
+        """
+        
+        self.full_data = BinnedData(data_yaml)
+        self.full_data.load_binned_data_from_hdf5(data_file)
+        self.estimated_bg = self.full_data.binned_data.project('Em', 'Phi', 'PsiChi')
```
[the remark this answers is not in the document]
Comment (ckarwin) [reply cosipy/background_estimation/ContinuumEstimation.py:92]: Done.

**Comment by ckarwin** [comment]
Thanks again for your review, @hiyoneda! It was very helpful. I implemented all your comments, and I also added unit tests. I think it can be merged if you think it's ok.

**Comment by hiyoneda** [comment]
@ckarwin Thanks. It looks good. I have two more comments on the notebook. After they are reflected, I think that it can be merged.
1. On the 3rd cell, the response file is downloaded, but the file is not consistent with the one we load later. Should it be `SMEXv12.Continuum.HEALPixO3_10bins_log_flat.binnedimaging.imagingresponse.nonsparse_nside8.area.good_chunks_unzip.earthocc.zip`?
2. At the last 2nd cell, the background file is generated, but only 1 Em bin and 2 Phi bins are considered. It is already mentioned, but can you add more information? Especially, how the values in other bins are filled (I suppose they are the same as the actually events), and how long it takes to run through all of bins roughly.

**Comment by ckarwin** [comment]
@hiyoneda Ok, it's been updated. 

1. Yes, the correct file is `SMEXv12.Continuum.HEALPixO3_10bins_log_flat.binnedimaging.imagingresponse.nonsparse_nside8.area.good_chunks_unzip.earthocc.zip`
2. I added the following text: 

> Note that the current code has not yet been optimized for speed, as it uses a simple nested for loop. The time required to generate the estimated background using all bins is roughyly 4 hours. The option to use a subset of the Em bins and/or Phi bins may be useful for analyses that also use a given subset, but at this point the main motivation for this option is for demonstrational purposes, and when using this option, nothing is done with the other bins.

**Review by hiyoneda — APPROVED** [review]
Thanks! It can be merged.

**Pull request merged**